analyses/hello-clusters/02_compare-clustering-parameters.Rmd

```diff
@@ -0,0 +1,466 @@
+stability_list <- cluster_results_list |>
+  purrr::map(
+    \(cluster_df) {
+      nn <- unique(cluster_df$nn)
+
+      # calculate stability, passing in the parameter value used for this iteration
+      rOpenScPCA::calculate_stability(pca_matrix, cluster_df, nn = nn)
+    }
+  )
+```
+
+We'll again combine the output of `stability_list` into a single data frame and plot `ari` values across `nn` parameterizations.
+
+
+```{r combine plot stability}
+stability_df <- dplyr::bind_rows(stability_list)
+
+ggplot(stability_df) +
+  aes(x = as.factor(nn), y = ari, fill = as.factor(nn)) +
+  geom_boxplot() +
+  scale_fill_brewer(palette = "Pastel2") +
+  labs(
+    x = "Number of nearest neighbors",
+    y = "Adjusted Rand Index"
+  ) +
+  theme(legend.position = "none")
+```
+
+Here, we see that a nearest neighbors value of 20 or 30 leads to more stable clustering results compared to 10. 
+
+
+## Varying multiple clustering parameters
+
+The previous section demonstrated how to calculate clusters and QC metrics when varying one parameter, but it is possible to vary multiple parameters at once with `rOpenScPCA::sweep_clusters()`.
+In this section, we'll show an overview of how you might write code to vary two parameters at once (here, nearest neighbors and resolution as examples) and visualize results.
+
+
+```{r sweep two parameters}
+# Define vectors of parameters to vary
+nn_values <- seq(10, 30, 10)
+res_values <- seq(5, 15, 5) / 10
+
+
+cluster_results_list <- rOpenScPCA::sweep_clusters(
+  pca_matrix,
+  nn = nn_values,
+  resolution = res_values
+)
+```
+
+This resulting list now has 9 different clustering results, for each combination of `nn_values` and `res_values`:
+
+
+```{r length cluster_results_list two parameters}
+length(cluster_results_list)
+```
+
+
+### Visualize clusters
+
+Next, we'll iterate over `cluster_results_list` to plot the UMAPs.
+This time, we'll use `purrr::map()` and pull out parameters from each iteration's `cluster_df` to form the UMAP panel title.
+
+```{r plot nn res umaps, fig.height = 14}
+umap_plots <- cluster_results_list |>
+  purrr::map(
+    \(cluster_df) {
+      # Add a column with cluster assignments to umap_df
+      umap_df_plot <- umap_df |>
+        dplyr::mutate(cluster = cluster_df$cluster)
+
+      # Create a title for the UMAP with both parameters
+      umap_title <- glue::glue(
+        "nn: {unique(cluster_df$nn)}; res: {unique(cluster_df$resolution)}"
+      )
+
+      # Plot the UMAP, colored by the new cluster variable
+      ggplot(umap_df_plot, aes(x = UMAP1, y = UMAP2, color = cluster)) +
+        geom_point(alpha = 0.6) +
+        labs(title = umap_title) +
+        # We'll add a couple UMAP-specific plot settings
+        coord_equal() +
+        theme(
+          axis.ticks = element_blank(),
+          axis.text = element_blank(),
+          legend.position = "bottom"
+        )
+    }
+  )
+
+# Print the plots with patchwork::wrap_plots()
+patchwork::wrap_plots(umap_plots)
```

Comment:
some of the plots have legends that are unfortunately cut off. Is that fixable somehow, I wonder? 

I would also add an explicit `ncol` to create the 3x3 grid.

```suggestion
patchwork::wrap_plots(umap_plots, ncol = 3)
```